vmm/src/lib.rs

```diff
@@ -1342,18 +1337,39 @@ impl Vmm {
 
         self.init_guest_memory()?;
 
-        self.setup_interrupt_controller()?;
+        let vcpus;
+
+        #[cfg(target_arch = "x86_64")]
+        {
+            self.setup_interrupt_controller()?;
+
+            self.attach_virtio_devices()?;
+            self.attach_legacy_devices()?;
+
+            let entry_addr = self.load_kernel()?;
+
+            self.configure_system()?;
 
-        self.attach_virtio_devices()?;
-        self.attach_legacy_devices()?;
+            self.register_events()?;
 
-        let entry_addr = self.load_kernel()?;
+            vcpus = self.create_vcpus(entry_addr, request_ts)?;
+        }
 
-        self.configure_system()?;
+        #[cfg(target_arch = "aarch64")]
```

Comment:
Maybe it's explained in the missing part of the commit message, but it's not clear why these two setup steps are so different. Is it an error on x86_64 to create the vcpus earlier? Why can't the first step be load_kernel on both?

Reply:
I updated the commit message and also put a comment before branching the implementation. Take a look.

Reply:
Looks much cleaner now. 

Could it be possible to further reduce duplication by doing the following:
```
            let entry_addr = self.load_kernel()?;

            #[cfg(target_arch = "aarch64")]
            vcpus = self.create_vcpus(entry_addr, request_ts)?;

            self.setup_interrupt_controller()?;
            self.attach_virtio_devices()?;
            self.attach_legacy_devices()?;

            #[cfg(target_arch = "x86_64")]
            vcpus = self.create_vcpus(entry_addr, request_ts)?;
```

Reply:
nit: Would be nice to also add a comment for both like:
```
// On aarch64 we need to create vcpus before creating the interrupt controller.
#[cfg(target_arch = "aarch64")]
vcpus = self.create_vcpus(entry_addr, request_ts)?;
...
// On x86_64 we need to create vcpus after setting up everything else.
#[cfg(target_arch = "x86_64")]
vcpus = self.create_vcpus(entry_addr, request_ts)?;
```

(Or smth similar)